### PR TITLE
Add proc macro package for automatic `IndexList<T>` implementation

### LINKED ISSUE #736: Provide proc macro package for automatic `IndexList<T>` implementation on any index struct
Currently, it would be kinda troublesome to use `IndexedMap` because we need to manually implement `IndexList` for every struct that we want to use as the index.

Rust's procedural macro feature enabled us to easily provide 1 line of code instead of manually typing the impl by ourselves for every struct we created.

Implementing this and reexport it in `cw-storage-plus` package (may be in an additional cfg features) will makes dev life much easier

Proposal: reduce this,

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct Token {
  pub owner: Addr,
  pub ticker: String,
  pub identifier: u8, // <---- unique value
}

pub struct TokenIndexes<'a> {
  pub identifier: UniqueIndex<'a, U8Key, Token>,
}

// IndexList is just boilerplate code for fetching a struct's indexes
impl<'a> IndexList<Token> for TokenIndexes<'a> {
  fn get_indexes(&'_ self) -> Box<dyn Iterator<Item=&'_ dyn Index<Token>> + '_> {
    let v: Vec<&dyn Index<Token>> = vec![&self.identifier];
    Box::new(v.into_iter())
  }
}
```

to this

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct Token {
  pub owner: Addr,
  pub ticker: String,
  pub identifier: u8, // <---- unique value
}

#[index_list(Token)]
pub struct TokenIndexes<'a> {
  pub identifier: UniqueIndex<'a, U8Key, Token>,
}
```

### PR DESCRIPTION
**Pull request by y-pakorn**
This will resolve #736

`macro` crate feature on `cw-storage-plus` is not enabled by default. 

`cw-storage-macro` doesn't need to be published into crate.io because the macro is re-exported in `cw-storage-plus` package.

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
pub struct Token {
  pub owner: Addr,
  pub ticker: String,
  pub identifier: u8,
}

#[index_list(Token)]
pub struct TokenIndexes<'a> {
  pub identifier: UniqueIndex<'a, U8Key, Token>,
}
```

### CODE REVIEW

File: packages/storage-macro/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "cw-storage-macro"
+version = "0.13.4"
+authors = ["yoisha <48324733+y-pakorn@users.noreply.github.com>"]
+edition = "2018"
+description = "Macro helper for storage package"
```
Comment (maurolacy) [inline packages/storage-macro/Cargo.toml:6]:
```suggestion
description = "Macro helpers for storage-plus"
```

File: packages/storage-macro/NOTICE [hunk 2]
```diff
@@ -0,0 +1,14 @@
+CW-Storage-Macro: Macro helper for storage package
+Copyright (C) 2020 Confio OÜ
```
Comment (maurolacy) [inline packages/storage-macro/NOTICE:2]:
```suggestion
CW-Storage-Macro: Macro helpers for storage-plus
Copyright (C) 2022 Confio GmbH
```

File: packages/storage-macro/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# CW-Storage-Plus: Macro helper for storage package
```
Comment (maurolacy) [inline packages/storage-macro/README.md:1]:
```suggestion
# CW-Storage-Plus: Macro helpers for storage-plus 
```

File: packages/storage-macro/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# CW-Storage-Plus: Macro helper for storage package 
+
+Procedural macros helper for interacting with cw-storage-plus and cosmwasm-storage.
+
+## Current features
+
+Auto generate IndexList impl for your indexes struct.
```
Comment (maurolacy) [inline packages/storage-macro/README.md:7]:
```suggestion
Auto generate an `IndexList` impl for your indexes struct.
```

File: packages/storage-macro/README.md [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+```rust
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+struct TestStruct {
+    id: u64,
+    id2: u32,
+    addr: Addr,
+}
+
+#[index_list(TestStruct)] // <- Add this line right here,.
```
Comment (maurolacy) [inline packages/storage-macro/README.md:17]:
```suggestion
#[index_list(TestStruct)] // <- Add this line right here.
```

File: packages/storage-macro/tests/index_list.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    #[index_list(TestStruct)]
+    struct TestIndexes<'a> {
+        id: MultiIndex<'a, u32, TestStruct, u64>,
+        addr: UniqueIndex<'a, Addr, TestStruct>,
+    }
+
+    let _: IndexedMap<u64, TestStruct, TestIndexes> = IndexedMap::new(
+        "t",
+        TestIndexes {
+            id: MultiIndex::new(|t| t.id2, "t", "t_2"),
```
Comment (maurolacy) [inline packages/storage-macro/tests/index_list.rs:24]:
```suggestion
            id: MultiIndex::new(|t| t.id2, "t", "t_id2"),
```

File: packages/storage-plus/Cargo.toml [hunk 7]
```diff
@@ -20,6 +21,7 @@ bench = false
 cosmwasm-std = { version = "1.0.0", default-features = false }
 schemars = "0.8.1"
 serde = { version = "1.0.103", default-features = false, features = ["derive"] }
+cw-storage-macro = { version = "0.13.4", optional = true, path = "../storage-macro" }
```
Comment (maurolacy) [inline packages/storage-plus/Cargo.toml:24]: Wouldn't this introduce circular deps? I guess `optional = true` breaks that, right?

**Review by maurolacy — APPROVED** [review]
Thanks for your contribution.

LTGM. Just some docs / comments style suggestions.

**Comment by maurolacy** [comment]
Seems the code coverage tool is being confused by the macro. We need to solve that before merging.

File: packages/storage-plus/Cargo.toml (hunk 7 above)
Comment (y-pakorn) [reply packages/storage-plus/Cargo.toml:24]: Afaik no, cw-storage-plus imported in cw-storage-macro package is only for tests as it is included in dev-dependencies.
https://github.com/CosmWasm/cw-plus/blob/3f29e20cf1f39edc924b62e0d0296cfdb61f5f1b/packages/storage-macro/Cargo.toml#L19

**Comment by y-pakorn** [comment]
> Seems the code coverage tool is being confused by the macro. We need to solve that before merging.

Might we need an additional tarpaulin config flag to cover integration tests?

**Comment by maurolacy** [comment]
> > Seems the code coverage tool is being confused by the macro. We need to solve that before merging.
> 
> Might we need an additional tarpaulin config flag to cover integration tests?

Not sure how to address this. Also, please add the new storage-macro package to the README.md, under **Utilities**.

**Comment by maurolacy** [comment]
> > > Seems the code coverage tool is being confused by the macro. We need to solve that before merging.
> > 
> > 
> > Might we need an additional tarpaulin config flag to cover integration tests?
> 
> Not sure how to address this. Also, please add the new storage-macro package to the README.md, under **Utilities**.

This is just because this is a new package, that doesn't have previous runs.

Merging now, as adding to README is not needed, as this is a private package.

**Pull request merged**